dpkg/scanner.go

```diff
@@ -0,0 +1,208 @@
+
+// Kind implements scanner.VersionedScanner.
+func (ps *Scanner) Kind() string { return kind }
+
+// Scan proxies the call to ScanContext.
+func (ps *Scanner) Scan(layer *claircore.Layer) ([]*claircore.Package, error) {
+	ctx := context.TODO()
+	ctx = log.Logger.WithContext(ctx)
+	return ps.ScanContext(ctx, layer)
+}
+
+// ScanContext attempts to find a dpkg database within the layer and read all of the
+// installed packages it can find in the "status" file.
+//
+// It's expected to return (nil, nil) if there's no dpkg database in the layer.
+//
+// It does not respect any dpkg configuration files.
+func (ps *Scanner) ScanContext(ctx context.Context, layer *claircore.Layer) ([]*claircore.Package, error) {
+	// Preamble
+	defer trace.StartRegion(ctx, "Scanner.Scan").End()
+	trace.Log(ctx, "layer:sha256", layer.Hash)
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "package_scanner").
+		Str("name", ps.Name()).
+		Str("version", ps.Version()).
+		Str("kind", ps.Kind()).
+		Str("layer", layer.Hash).
+		Logger()
+	log.Debug().Msg("start")
+	defer log.Debug().Msg("done")
+
+	// Grab a handle to the tarball, make sure we can seek.
+	// If we can't, we'd need another reader for every database found.
+	// It's cleaner to just demand that it's a seeker.
+	rd, err := layer.Reader()
+	if err != nil {
+		log.Warn().Err(err).Msg("opening layer failed")
+		return nil, fmt.Errorf("opening layer failed: %w", err)
+	}
+	defer rd.Close()
+	r, ok := rd.(io.ReadSeeker)
+	if !ok {
+		err := errors.New("unable to coerce to io.Seeker")
+		log.Warn().Err(err).Msg("opening layer failed")
+		return nil, fmt.Errorf("opening layer failed: %w", err)
+	}
+
+	tr := tar.NewReader(r)
+	// This is a map keyed by directory. A "score" of 2 means this is almost
+	// certainly a dpkg database.
+	loc := make(map[string]int)
+Find:
+	for {
+		h, err := tr.Next()
+		switch err {
+		case nil:
+		case io.EOF:
+			break Find
+		default:
+			log.Warn().Err(err).Msg("reading next header failed")
+			return nil, fmt.Errorf("reading next header failed: %w", err)
+		}
+		switch filepath.Base(h.Name) {
+		case "status", "available":
+			if h.Typeflag == tar.TypeReg {
+				loc[filepath.Dir(h.Name)]++
+			}
+		}
+	}
+	log.Debug().Msg("scanned for possible databases")
+
+	// If we didn't find anything, this loop is completely skipped.
+	var pkgs []*claircore.Package
+	for p, x := range loc {
+		if x != 2 { // If we didn't find both files, skip this directory.
+			continue
+		}
+		log.Debug().Str("database", p).Msg("examining package database")
+
+		// Reset the tar reader.
+		if n, err := r.Seek(0, io.SeekStart); n != 0 || err != nil {
+			log.Warn().Err(err).Msg("unable to seek reader")
+			return nil, fmt.Errorf("unable to seek reader: %w", err)
+		}
+		tr = tar.NewReader(r)
+
+		// We want the "status" file, so search the archive for it.
+		fn := filepath.Join(p, "status")
+		var db io.Reader
+		var h *tar.Header
+		for h, err = tr.Next(); err == nil; h, err = tr.Next() {
```

Comment:
almost certainly not worth it as far as optimization but for a conversation, you already seeked the file to the header to find the status file. I believe you could use that offset and record it in your map, then seek the reader directly to that file's header.